Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs

```diff
@@ -1897,18 +1901,26 @@ private void AddStaticFileDependencyToFileAccessManifest(FileArtifact dependency
             // Note that they would perhaps fail otherwise (simplifies the observed access checks in the first place).
 
             var path = dependency.Path;
+
+            bool pathIsUnderSharedOpaque = TryGetContainingSharedOpaqueRoot(path, getOutmostRoot: true, out var sharedOpaqueRoot);
+
             m_fileAccessManifest.AddPath(
                 path,
                 values: FileAccessPolicy.AllowRead | FileAccessPolicy.AllowReadIfNonexistent,
-                mask: m_excludeReportAccessMask & ~FileAccessPolicy.AllowRealInputTimestamps); // Make sure we fake the input timestamp
+                // Make sure we fake the input timestamp
+                // The file dependency may be under the cone of a shared opaque, which will give write access
+                // to it. Explicitly block this, since we want inputs to not be written. Observe we already know 
+                // this is not a rewrite.
+                mask: m_excludeReportAccessMask & ~FileAccessPolicy.AllowRealInputTimestamps 
+                    & (pathIsUnderSharedOpaque ? ~FileAccessPolicy.AllowWrite: FileAccessPolicy.MaskNothing)); 

```

Comment:
Would it be ok to unconditionally add `~FileAccessPolicy.AllowWrite` for all static dependencies, regardless of whether they are under a shared opaque directory or not.

Reply:
Tried that first. The problem is that there can be inputs that are also untracked (e.g. a source is declared, but since it is overwritten, the output - which is the same path as the input - gets flagged as untracked). And in the way the manifest is constructed, one can only add masks but not remove them. So after adding ~Write here, there is no direct way to remove that flag in case that same path happens to be untracked.

---
In reply to: [284862416](https://github.com/microsoft/BuildXL/pull/316#discussion_r284862416) [](ancestors = 284862416)

Reply:
thanks